Core/AssetComparer.cs

```
@@ -0,0 +1,86 @@
+
+                UpdateAssets(intersectingAssets);
+            }
+
+            List<T> addedAssets = GetAddedAssets(assetsFromList, intersectingAssets);
+            
+            if (addedAssets.Count > 0)
+            {
+                NewAssetsFound?.Invoke(addedAssets);
+            }
+        }
+
+        /// <summary>
+        /// Gets the intersecting assets between the current assets and the assets from the list.
+        /// </summary>
+        /// <param name="assetsFromList">The assets to compare to the current assets.</param>
+        /// <returns>The intersecting assets between the current assets and the assets from the list.</returns>
+        private IEnumerable<T> GetIntersectingAssets(IEnumerable<T> assetsFromList)
+        {
+            return assetsFromList
+               .Where(newAsset => currentAssets
+                    .Any(asset => newAsset.AssetId == asset.AssetId))
+               .ToList();
+        }
+
+        private void UpdateAssets(IEnumerable<T> intersectingAssets)
+        {
+            UpdateAssetStates(intersectingAssets);
+            UpdateAssetHolders(intersectingAssets);
+        }
+        
+        private void UpdateAssetStates(IEnumerable<T> intersectingAssets)
+        {
+            foreach (T currentAsset in currentAssets)
+            {
+                if (!intersectingAssets.Any(asset => asset.AssetId == currentAsset.AssetId))
+                {
+                    currentAsset.SetState(AssetState.Missing);
+                }
+            }
+        }
+
+        private void UpdateAssetHolders(IEnumerable<T> intersectingAssets)
+        {
+            foreach (T newAsset in intersectingAssets)
+            {
+                T currentAsset = currentAssets.Find(oldAsset => oldAsset.AssetId == newAsset.AssetId);
+                
+                if (currentAsset.CurrentAssetHolder == null || !currentAsset.CurrentAssetHolder.Equals(newAsset.CurrentAssetHolder))
```

CurrentAssetHolder should never be null - If null it should throw an exception.